diff --git a/config-init/config.inc.php b/config-init/config.inc.php
index 295da83..0031578 100644
--- a/config-init/config.inc.php
+++ b/config-init/config.inc.php
@@ -40,12 +40,19 @@ if (!headers_sent())
 	header('Content-Type: text/html; charset=utf-8');
 
 /* No settings file? goto installer... */
-if (!file_exists(_PS_ROOT_DIR_.'/config/settings.inc.php'))
+if (!file_exists(_PS_ROOT_DIR_.'/config/settings.inc.php') || file_exists(_PS_ROOT_DIR_.'/config/install'))
 {
+	/* Fix to override default database credentials using settings_platform.inc.php. */
+	if(file_exists(_PS_ROOT_DIR_.'/config/install')) {
+		copy(_PS_ROOT_DIR_.'/config/settings_platform.inc.php', _PS_ROOT_DIR_.'/config/settings.inc.php');
+		/* Remove the 'install' file. */
+		unlink(_PS_ROOT_DIR_.'/config/install');
+	}
 	if (file_exists(dirname(__FILE__).'/../install'))
 		header('Location: install/');
 	elseif (file_exists(dirname(__FILE__).'/../install-dev'))
-		header('Location: install-dev/');
+		/* Fix to prevent infinite loop on "install-dev/". */
+		header('Location: /install-dev/index.php');
 	else
 		die('Error: "install" directory is missing');
 	exit;
